File: clients/python/examples/test_basic.py
"""test_basic.py — end-to-end example for the ScrivaDB Python client.

Prerequisites:
    - ScrivaDB server running: `make run` from the repo root.
    - Client installed:      `pip install .` from clients/python.

Run:
    python examples/test_basic.py
"""

from scriva import AlreadyExistsError, ScrivaDB, NotFoundError


def main() -> None:
    db = ScrivaDB("localhost", 5433, "dev-key")

    # --- Collection management ---
    print("=== Collections ===")
    db.create_collection("test_py")
    print("Created collection. All collections:", db.list_collections())

    # --- Insert ---
    print("\n=== Insert ===")
    id1 = db.insert("test_py", {"name": "Alice", "age": 30, "role": "admin"})
    id2 = db.insert("test_py", {"name": "Bob", "age": 25, "role": "user"})
    id3 = db.insert("test_py", {"name": "Carol", "age": 35, "role": "admin"})
    print("Inserted IDs:", id1, id2, id3)

    ids = db.insert_many(
        "test_py",
        [
            {"name": "Dave", "age": 28, "role": "user"},
            {"name": "Eve", "age": 22, "role": "user"},
        ],
    )
    print("InsertMany IDs:", ids)

    # --- Find by ID ---
    print("\n=== FindById ===")
    print("Record:", db.find_by_id("test_py", id1))

    # --- Find with filter ---
    print("\n=== Find (filter: role=admin) ===")
    admins = db.find(
        "test_py", {"field": "role", "op": "eq", "value": "admin"}, order_by="name"
    )
    print("Admins:", [f"{r['id']}: {r['data']}" for r in admins])

    # --- Projection (N2): only return selected fields ---
    print("\n=== Find with projection (fields=['name']) ===")
    projected = db.find("test_py", fields=["name"])
    print("Projected:", [r["data"] for r in projected])

    # --- Multi-field sort (N3) ---
    print("\n=== Find (multi-field order_by: role asc, age desc) ===")
    sorted_recs = db.find(
        "test_py", order_by=[("role", False), ("age", True)]
    )
    print("Sorted:", [(r["data"].get("role"), r["data"].get("age")) for r in sorted_recs])

    # --- Keyset pagination (N3) ---
    print("\n=== Keyset pagination (limit=2, order_by=age) ===")
    page, token = db.find_page("test_py", limit=2, order_by="age")
    print("Page 1:", [r["data"].get("name") for r in page])
    while token:
        page, token = db.find_page("test_py", limit=2, order_by="age", page_token=token)
        print("Next page:", [r["data"].get("name") for r in page])

    # --- AND filter ---
    print("\n=== Find (AND: role=user AND age>=25) ===")
    filtered = db.find(
        "test_py",
        {
            "and": [
                {"field": "role", "op": "eq", "value": "user"},
                {"field": "age", "op": "gte", "value": "25"},
            ]
        },
    )
    print("Filtered:", [r["data"] for r in filtered])

    # --- Find with limit ---
    print("\n=== Find (limit 2) ===")
    for r in db.find("test_py", limit=2):
        print(" -", r["id"], r["data"])

    # --- Update ---
    print("\n=== Update ===")
    db.update("test_py", id1, {"name": "Alice", "age": 31, "role": "superadmin"})
    print("Updated:", db.find_by_id("test_py", id1))

    # --- Delete ---
    print("\n=== Delete ===")
    print("Deleted id2:", db.delete("test_py", id2))

    # --- Indexes ---
    print("\n=== Indexes ===")
    db.ensure_index("test_py", "role")
    print("Indexes:", db.list_indexes("test_py"))

    indexed = db.find("test_py", {"field": "role", "op": "eq", "value": "user"})
    print("Users (via index):", [r["data"] for r in indexed])

    db.drop_index("test_py", "role")
    print("Indexes after drop:", db.list_indexes("test_py"))

    # --- Transactions ---
    print("\n=== Transactions ===")
    tx_id = db.begin_tx("test_py")
    print("TX ID:", tx_id)
    print("Committed:", db.commit_tx(tx_id))

    tx_id2 = db.begin_tx("test_py")
    print("Rolled back:", db.rollback_tx(tx_id2))

    # --- Keyed CRUD, Upsert & CAS (N1) ---
    print("\n=== Keyed CRUD (N1) ===")
    rec = db.upsert("test_py", "user:alice", {"name": "Alice", "score": 10})
    print("Upsert (insert):", rec)  # rev == 1
    rec = db.upsert("test_py", "user:alice", {"name": "Alice", "score": 20})
    print("Upsert (replace):", rec)  # rev == 2

    print("FindByKey:", db.find_by_key("test_py", "user:alice"))

    updated = db.update_by_key("test_py", "user:alice", {"name": "Alice", "score": 30})
    print("UpdateByKey:", updated)  # rev == 3

    # Keyed create — a duplicate key is rejected with AlreadyExistsError.
    kid = db.insert("test_py", {"name": "Frank"}, key="user:frank")
    print("Keyed insert id:", kid)
    try:
        db.insert("test_py", {"name": "Frank II"}, key="user:frank")
    except AlreadyExistsError as e:
        print("Duplicate key rejected as expected:", type(e).__name__)

    # Compare-and-swap on rev.
    current_rev = db.find_by_key("test_py", "user:alice")["rev"]
    cas = db.update_if_rev(
        "test_py", "user:alice", current_rev, {"name": "Alice", "score": 40}
    )
    print("CAS (fresh rev) swapped:", cas["swapped"], "-> rev", cas["record"]["rev"])
    stale = db.update_if_rev(
        "test_py", "user:alice", current_rev, {"name": "Alice", "score": 99}
    )
    print("CAS (stale rev) swapped:", stale["swapped"])  # False, clean no-op

    print("DeleteByKey:", db.delete_by_key("test_py", "user:frank"))
    try:
        db.find_by_key("test_py", "user:frank")
    except NotFoundError as e:
        print("Missing key raises as expected:", type(e).__name__)

    # --- Aggregations (N4) ---
    print("\n=== Aggregations (N4) ===")
    print("Total count:", db.count("test_py"))
    print(
        "Admin count:",
        db.count("test_py", {"field": "role", "op": "eq", "value": "admin"}),
    )
    by_role = db.group_by(
        "test_py", "role", aggregations=["sum", "avg", "min", "max"], metric="age"
    )
    for g in by_role:
        print(f"  role={g['group']!r} count={g['count']} numeric={g['numeric']}", end="")
        if g["numeric"]:
            print(f" sum={g['sum']} avg={g['avg']:.1f} min={g['min']} max={g['max']}")
        else:
            print()

    # --- Stats ---
    print("\n=== Stats ===")
    print("Stats:", db.stats("test_py"))

    # --- TTL (per-record and per-collection default) ---
    print("\n=== TTL ===")
    db.create_collection("test_py_ttl", default_ttl_seconds=3600)
    db.insert("test_py_ttl", {"kind": "inherits-collection-default"})
    db.insert("test_py_ttl", {"kind": "own-ttl"}, ttl_seconds=60)
    print("TTL collection stats:", db.stats("test_py_ttl"))
    db.drop_collection("test_py_ttl")

    # --- Maintenance ---
    print("\n=== Compact ===")
    print("Compacted:", db.compact("test_py"))

    # --- Backup ---
    print("\n=== Snapshot ===")
    n = db.snapshot_to_file("scriva-backup.tar.gz")
    print(f"Wrote {n} bytes to scriva-backup.tar.gz (restore with: tar xzf ...)")

    # --- Cleanup ---
    print("\n=== Cleanup ===")
    db.drop_collection("test_py")
    print("Collections after drop:", db.list_collections())

    db.close()
    print("\nAll done!")


if __name__ == "__main__":
    main()

File: clients/js/examples/test_basic.ts
/**
 * test_basic.ts — end-to-end example for the FileDB v2 TypeScript client.
 *
 * Prerequisites:
 *   - FileDB server running: `make run` from the repo root
 *
 * Run:
 *   npx ts-node examples/test_basic.ts
 */

import { FileDB } from '../src';

async function main(): Promise<void> {
  const db = new FileDB('localhost', 5433, 'dev-key');

  // --- Collection management ---
  console.log('=== Collections ===');
  await db.createCollection('test_js');
  console.log('Created collection. All collections:', await db.listCollections());

  // --- Insert ---
  console.log('\n=== Insert ===');
  const id1 = await db.insert('test_js', { name: 'Alice', age: 30, role: 'admin' });
  const id2 = await db.insert('test_js', { name: 'Bob',   age: 25, role: 'user'  });
  const id3 = await db.insert('test_js', { name: 'Carol', age: 35, role: 'admin' });
  console.log('Inserted IDs:', id1, id2, id3);

  const ids = await db.insertMany('test_js', [
    { name: 'Dave', age: 28, role: 'user' },
    { name: 'Eve',  age: 22, role: 'user' },
  ]);
  console.log('InsertMany IDs:', ids);

  // --- Find by ID ---
  console.log('\n=== FindById ===');
  const record = await db.findById('test_js', id1);
  console.log('Record:', record);

  // --- Find with filter + multi-field sort (N3) ---
  console.log('\n=== Find (filter: role=admin, order by name) ===');
  const admins = await db.findAll('test_js', {
    filter: { field: 'role', op: 'eq', value: 'admin' },
    orderBy: [{ field: 'name' }],
  });
  console.log('Admins:', admins.map(r => `${r.id}: ${JSON.stringify(r.data)}`));

  // --- AND filter ---
  console.log('\n=== Find (AND: role=user AND age>=25) ===');
  const filtered = await db.findAll('test_js', {
    filter: {
      and: [
        { field: 'role', op: 'eq',  value: 'user' },
        { field: 'age',  op: 'gte', value: '25'   },
      ],
    },
  });
  console.log('Filtered:', filtered.map(r => r.data));

  // --- Streaming find ---
  console.log('\n=== Find (streaming, limit 2) ===');
  for await (const r of db.find('test_js', { limit: 2 })) {
    console.log(' -', r.id, r.data);
  }

  // --- Projection (N2): only return selected fields ---
  console.log('\n=== Find (projection: name only) ===');
  const projected = await db.findAll('test_js', {
    fields: ['name'],
    orderBy: [{ field: 'name' }],
  });
  console.log('Projected (id/key/rev always included):', projected.map(r => r.data));

  // --- Keyset pagination (N3): page through by cursor ---
  console.log('\n=== Find (keyset pagination, page size 2) ===');
  let token = '';
  let page = 0;
  do {
    const p = await db.findPage('test_js', {
      orderBy: [{ field: 'age' }],
      limit: 2,
      pageToken: token,
    });
    console.log(` page ${++page}:`, p.records.map(r => `${r.data.name}(${r.data.age})`));
    token = p.pageToken;
  } while (token);

  // --- Keyed CRUD / Upsert / CAS (N1) ---
  console.log('\n=== Keyed CRUD (N1) ===');
  const up1 = await db.upsert('test_js', 'user:alice', { name: 'Alice', tier: 'free' });
  console.log('Upsert insert:', up1.key, 'rev', up1.rev, up1.data);
  const up2 = await db.upsert('test_js', 'user:alice', { name: 'Alice', tier: 'pro' });
  console.log('Upsert replace:', up2.key, 'rev', up2.rev, up2.data);

  const byKey = await db.findByKey('test_js', 'user:alice');
  console.log('FindByKey:', byKey?.data, 'rev', byKey?.rev);
  console.log('FindByKey (missing → null):', await db.findByKey('test_js', 'user:nobody'));

  const w = await db.updateByKey('test_js', 'user:alice', { name: 'Alice', tier: 'enterprise' });
  console.log('UpdateByKey → rev', w.rev);

  // Compare-and-swap: succeeds with the current rev, no-ops with a stale one.
  const cur = await db.findByKey('test_js', 'user:alice');
  const ok = await db.updateIfRev('test_js', 'user:alice', cur!.rev, { name: 'Alice', tier: 'vip' });
  console.log('CAS with current rev → swapped:', ok.swapped, 'newRev', ok.record?.rev);
  const stale = await db.updateIfRev('test_js', 'user:alice', cur!.rev, { name: 'Alice', tier: 'x' });
  console.log('CAS with stale rev → swapped:', stale.swapped);

  console.log('DeleteByKey:', await db.deleteByKey('test_js', 'user:alice'));
  console.log('DeleteByKey (missing → false):', await db.deleteByKey('test_js', 'user:nobody'));

  // --- Aggregations (N4) ---
  console.log('\n=== Aggregations (N4) ===');
  console.log('Total count:', await db.count('test_js'));
  console.log('Count (role=user):', await db.count('test_js', { field: 'role', op: 'eq', value: 'user' }));
  const byRole = await db.groupBy('test_js', 'role', {
    field: 'age',
    aggregations: ['sum', 'avg', 'min', 'max'],
  });
  for (const g of byRole) {
    console.log(` group ${JSON.stringify(g.group)}: count=${g.count}` +
      (g.numeric ? ` sum=${g.sum} avg=${g.avg} min=${g.min} max=${g.max}` : ''));
  }

  // --- Update ---
  console.log('\n=== Update ===');
  await db.update('test_js', id1, { name: 'Alice', age: 31, role: 'superadmin' });
  console.log('Updated:', await db.findById('test_js', id1));

  // --- Delete ---
  console.log('\n=== Delete ===');
  const deleted = await db.delete('test_js', id2);
  console.log('Deleted id2:', deleted);

  // --- Indexes ---
  console.log('\n=== Indexes ===');
  await db.ensureIndex('test_js', 'role');
  console.log('Indexes:', await db.listIndexes('test_js'));

  // find via index (single eq on indexed field)
  const indexed = await db.findAll('test_js', {
    filter: { field: 'role', op: 'eq', value: 'user' },
  });
  console.log('Users (via index):', indexed.map(r => r.data));

  await db.dropIndex('test_js', 'role');
  console.log('Indexes after drop:', await db.listIndexes('test_js'));

  // --- Transactions ---
  console.log('\n=== Transactions ===');
  const txId = await db.beginTx('test_js');
  console.log('TX ID:', txId);
  const committed = await db.commitTx(txId);
  console.log('Committed:', committed);

  const txId2 = await db.beginTx('test_js');
  const rolledBack = await db.rollbackTx(txId2);
  console.log('Rolled back:', rolledBack);

  // --- Stats ---
  console.log('\n=== Stats ===');
  const s = await db.stats('test_js');
  console.log('Stats:', s);

  // --- TTL (per-record and per-collection default) ---
  console.log('\n=== TTL ===');
  await db.createCollection('test_js_ttl', 3600);
  await db.insert('test_js_ttl', { kind: 'inherits-collection-default' });
  await db.insert('test_js_ttl', { kind: 'own-ttl' }, 60);
  console.log('TTL collection stats:', await db.stats('test_js_ttl'));
  await db.dropCollection('test_js_ttl');

  // --- Maintenance ---
  console.log('\n=== Compact ===');
  console.log('Compacted:', await db.compact('test_js'));

  // --- Backup ---
  console.log('\n=== Snapshot ===');
  const bytes = await db.snapshotToFile('filedb-backup.tar.gz');
  console.log(`Wrote ${bytes} bytes to filedb-backup.tar.gz (restore with: tar xzf ...)`);

  // --- Cleanup ---
  console.log('\n=== Cleanup ===');
  await db.dropCollection('test_js');
  console.log('Collections after drop:', await db.listCollections());

  db.close();
  console.log('\nAll done!');
}

main().catch((err: unknown) => {
  console.error(err);
  process.exit(1);
});
